x-pack/solutions/security/plugins/security_solution/public/asset_inventory/hooks/use_dynamic_entity_flyout.ts

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useExpandableFlyoutApi } from '@kbn/expandable-flyout';
+import type { UniversalEntityEcs } from '@kbn/securitysolution-ecs/src/universal_entity';
+import {
+  ASSET_INVENTORY_EXPAND_FLYOUT_SUCCESS,
+  ASSET_INVENTORY_EXPAND_FLYOUT_ERROR,
+  uiMetricService,
+} from '@kbn/cloud-security-posture-common/utils/ui_metrics';
+import { METRIC_TYPE } from '@kbn/analytics';
+import { i18n } from '@kbn/i18n';
+import { useKibana } from '../../common/lib/kibana';
+import {
+  HostPanelKey,
+  ServicePanelKey,
+  UniversalEntityPanelKey,
+  UserPanelKey,
+} from '../../flyout/entity_details/shared/constants';
+import { useOnExpandableFlyoutClose } from '../../flyout/shared/hooks/use_on_expandable_flyout_close';
+
+interface InventoryFlyoutProps {
+  entity: UniversalEntityEcs;
+  scopeId?: string;
+  contextId?: string;
+}
+
+interface SecurityFlyoutPanelsCommonParams {
+  scopeId?: string;
+  contextId?: string;
+  [key: string]: unknown;
+}
+
+type FlyoutParams =
+  | {
+      id: typeof UniversalEntityPanelKey;
+      params: { entity: InventoryFlyoutProps['entity'] };
+    }
+  | { id: typeof UserPanelKey; params: { userName: string } & SecurityFlyoutPanelsCommonParams }
+  | { id: typeof HostPanelKey; params: { hostName: string } & SecurityFlyoutPanelsCommonParams }
+  | {
+      id: typeof ServicePanelKey;
+      params: { serviceName: string } & SecurityFlyoutPanelsCommonParams;
+    };
+
+const getFlyoutParamsByEntity = (
+  entity: InventoryFlyoutProps['entity'],
+  scopeId?: string,
+  contextId?: string
+): FlyoutParams => {
+  const entitiesFlyoutParams: Record<InventoryFlyoutProps['entity']['type'], FlyoutParams> = {
+    universal: { id: UniversalEntityPanelKey, params: { entity } },
+    user: { id: UserPanelKey, params: { userName: entity.name, scopeId, contextId } },
+    host: { id: HostPanelKey, params: { hostName: entity.name, scopeId, contextId } },
+    service: { id: ServicePanelKey, params: { serviceName: entity.name, scopeId, contextId } },
+  } as const;
+
+  return entitiesFlyoutParams[entity.type];
+};
+
+export const useDynamicEntityFlyout = ({ onFlyoutClose }: { onFlyoutClose: () => void }) => {
+  const { openFlyout, closeFlyout } = useExpandableFlyoutApi();
+  const { notifications } = useKibana().services;
+  useOnExpandableFlyoutClose({ callback: onFlyoutClose });
+
+  const openDynamicFlyout = ({ entity, scopeId, contextId }: InventoryFlyoutProps) => {
+    const entityFlyoutParams = getFlyoutParamsByEntity(entity, scopeId, contextId);
+
+    // User, Host, and Service entity flyouts rely on entity name to fetch required data
+    if (entity.type !== 'universal' && !entity.name) {
```

Comment:
```suggestion
    if (entity.type !== UNIVERSAL_ENTITY_TYPE_UNIVERSAL && !entity.name) {
```
Not sure if this needs a constant right now, mainly because I would prefer it to be coming from ECS and not some local value